src/storage/mod.rs

```diff
@@ -0,0 +1,566 @@
+//! # dioxus-storage
+//! A library for handling local storage ergonomically in Dioxus
+//! ## Usage
+//! ```rust
+//! use dioxus_storage::use_storage;
+//! use dioxus::prelude::*;
+//! fn main() {
+//!     dioxus_web::launch(app)
+//! }
+//!
+//! fn app(cx: Scope) -> Element {
+//!     let num = use_persistent(cx, "count", || 0);
+//!     cx.render(rsx! {
+//!         div {
+//!             button {
+//!                 onclick: move |_| {
+//!                     num.modify(|num| *num += 1);
+//!                 },
+//!                 "Increment"
+//!             }
+//!             div {
+//!                 "{*num.read()}"
+//!             }
+//!         }
+//!     })
+//! }
+//! ```
+
+mod client_storage;
+mod persistence;
+
+pub use client_storage::{LocalStorage, SessionStorage};
+pub use persistence::{use_persistent, use_singleton_persistent};
+
+use dioxus::prelude::{current_scope_id, to_owned, ScopeState};
+use dioxus_signals::{Effect, Signal};
+use postcard::to_allocvec;
+use serde::{de::DeserializeOwned, Serialize};
+use std::any::Any;
+use std::fmt::{Debug, Display};
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+use tokio::sync::watch::error::SendError;
+use tokio::sync::watch::{Receiver, Sender};
+
+#[cfg(not(target_family = "wasm"))]
+pub use client_storage::{set_dir, set_dir_name, set_directory};
+
+/// A storage hook that can be used to store data that will persist across application reloads.
+///
+/// This hook returns a Signal that can be used to read and modify the state.
+pub fn use_storage<S, T>(cx: &ScopeState, key: S::Key, init: impl FnOnce() -> T) -> Signal<T>
```

Comment:
these hooks should have their own examples provided to show that they require passing a specific Storagebacking impl